### ISSUE TITLE
[Lattice API] Inconsistencies

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#795]
cc @attila-i-szabo 

As @chrisrothUT needs to check if a Lattice has a point group, we should add a way to make this easy to check.

Also, I noticed such method [`Lattice.point_group`](https://github.com/netket/netket/blob/cd70d32bc4d149cbb22d5331e0f4dec4606b0431/netket/graph/lattice.py#L607) where we have 
```python
    def point_group(self, point_group: Optional[PointGroup] = None) -> PermutationGroup:
        """
        Returns the action of `point_group` on the sites of `self` as a
        `PermutationGroup`. If no `point_group` is specified, uses the point group
        provided upon construction.
        """
        return self.space_group_builder(point_group).point_group
```

I don't understand what goes on in there? If you pass a point group... why would you be calling the method?
And why do we pass through the builder? if the class already has a point group... what happens?

**Comment by PhilipVinc** [comment]
Is this because `PointGroup` is an abstract object, and `point_group` is a permutation?

**Comment by attila-i-szabo** [comment]
Yep. The model is that you supply a point group in abstracto and the `Lattice` knows how to turn that into a `PermutationGroup`. The same model is used for generating the space group. That `Lattice` can have a `PointGroup` member is just for convenience, so `Square`, `Pyrochlore`, etc. can come preloaded with a sensible point group, for a generic lattice, the user has to know what makes sense.

> And why do we pass through the builder?

Because you preferred having this logic separate from `Lattice`, so it lives in the `SpaceGroupBuilder` class.

> if the class already has a point group... what happens?

`Lattice.space_group_builder` calls the `SpaceGroupBuilder` constructor with the `PointGroup` given to it, or if it's `None`, with `Lattice._point_group`.

**Comment by attila-i-szabo** [comment]
> As @chrisrothUT needs to check if a Lattice has a point group, we should add a way to make this easy to check.

Rather, the GCNN constructor should offer a way to change/supply the point group, because it might not be supplied with the Lattice, or the user wants to use something other than the default, and both of those are legit use cases

**Comment by femtobit** [comment]
The only potentially confusing thing I see here is that the function (returning `PermutationGroup`) its argument (of type `PointGroup`) have the exact same name `point_group`; I do understand why a `point_group -> point_group` mapping looks strange at first.

OTOH, I don't see a good way to rename one of this without making the names too long (e.g., `abstract_point_group: PointGroup` as argument). Not sure if that'd be worth it.

**Comment by attila-i-szabo** [comment]
1. For a generic lattice, you don't know what the point group is, unless the user specifies it by hand. I don't feel like forcing the user to specify one if they don't want to use this machinery.
2. We'd need to rewrite every specialised Lattice constructor to allow overriding the default point group.

**Comment by PhilipVinc** [comment]
I don't think 1 is an issue? I'm not arguing for always having a point group.
I',m just arguing for making it so that once you set it, you can't change it (unless you create another lattice).

It's also because the point group, as far as I can tell, is deeply linked to the lattice itself. You cannot use it without a lattice.
Or am I wrong?

For 2 we could simply have a way to copy it overriding the point group.

```
class Lattice():
   ...

   def copy(point_group=Unspecified):
        ...

```

**Comment by attila-i-szabo** [comment]
3. Models may depend on a particular point group down to the order the elements are listed in (I'm working on one where this is convenient). Now I can make sure I use a particular order on the level of the model constructor, instead of having to deal with this when constructing the lattice.

> It's also because the point group, as far as I can tell, is deeply linked to the lattice itself. You cannot use it without a lattice.
> Or am I wrong?

RIght now that's the only use case built into Netket. I wouldn't be so sure there won't be others, especially in user code

> ```
> class Lattice():
>    ...
> 
>    def copy(point_group=Unspecified):
>         ...
> ```

I wouldn't mind this if we got something in exchange for losing flexiblity. Say, we make sure `Lattice` is a netket dataclass, so the various group objects can be cached (there can be a lot of computation going into these). I don't care enough about making things pythonic for the sake of it...

**Comment by femtobit** [comment]
> Say, we make sure Lattice is a netket dataclass, so the various group objects can be cached (there can be a lot of computation going into these)

But that is independent of whether the point group is a member of lattice, isn't it? If it is created externally, it is only created once just by virtue of users typically only creating it once (and caching properties of the group itself is done there).

> For 2 we could simply have a way to copy it overriding the point group.

I do not see how this is less complicated then just having the point group be external to the lattice as it is now. Really, the only things that is kind-of at odds with the rest of the API is that some lattices have a default point group. (And maybe the name of the method.)

**Comment by attila-i-szabo** [comment]
> But that is independent of whether the point group is a member of lattice, isn't it? If it is created externally, it is only created once just by virtue of users typically only calling `Lattice.point_group` once for a given group (and caching properties of the group itself is done in that object).

I'm talking about the space groups that get generated by the `Lattice`. Calculating their product tables etc. takes quite a bit of time, so it would be nice if you could return the same `PermutationGroup` object (which caches all this) every time. It's straightforward to cache it if you fix the point group (in fact `SpaceGroupBuilder` is a dataclass, so on that level it's happening), but even if you supply the same `PointGroup` externally several times, you generate a new `SpaceGroupBuilder`. The situation would be better if the user were forced through the `SpaceGroupBuilder` interface, but we chose not to do that, except for fancy stuff like space group irreps.

> Really, the only things that is kind-of at odds with the rest of the API is that some lattices have a default point group.

Perhaps, but it makes the specialised constructors much more useful (in the sense that all the symmetry stuff works out of the box). There might be a more elegant way of dealing with this, but I wouldn't like to lose this functionality.

**Comment by PhilipVinc** [comment]
> Perhaps, but it makes the specialised constructors much more useful (in the sense that all the symmetry stuff works out of the box). There might be a more elegant way of dealing with this, but I wouldn't like to lose this functionality.

I agree with @femtobit . What feels weird in this interface is that you can have a default point group but also supply an external one.
I see the reasoning for having a 'default' one provided with standard lattices. It makes perfect sense, and I agree that we should't lose such feature.
At the same time, we have two code paths with different failure points: we have Lattices with and Lattices without a point group. 
When you build a GCNN you may have both, but you can also pass an overrride. Which is needed if the lattice does not have a point group.

This logic is complex. Has many cases and therefore it seems to me we can simplify it. In general I find that if there is one well defined way to do something it's better.
For example, i find it confusing that to override the point group of a lattice passed to a GCNN, i need to pass the lattice and the point group, but i can also build a lattice with the right point group.

That's the reason i was proposing making it so that the point group is always inside a lattice, and make it easy to change the point group inside a lattice. 
This would make all our logic easier.

**Comment by PhilipVinc** [comment]
This is an orthogonal issue wrt making our Lattice a dataclass (which, by the way, would be relatively simple to do, but I don't think we really need that right now).
It's also orthogonal to having a pythonic interface. 
Half this stuff is functional, which arguably is not pythonic at all.

This is all about having a clean API that is easier to maintain in the future.

**Comment by femtobit** [comment]
@PhilipVinc 
> At the same time, we have two code paths with different failure points: we have Lattices with and Lattices without a point group.
> When you build a GCNN you may have both, but you can also pass an overrride. Which is needed if the lattice does not have a point group.

Wanting to change this by changing the lattice API is maybe not actually the right place: The fact that the GCNN takes the default point group of a lattice is only a slight convenience. Correct me if I remember the issue here wrong, but woudn't it be a clean solution to require an explicit point group to be used for creating the `GCNN`? Then, the distinction between a lattice having a default point group or not is no longer relevant for the GCNN.
